src/raftstore/store/store.rs

```diff
@@ -1134,15 +1285,21 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         let mut ready_results = Vec::with_capacity(append_res.len());
         for (mut ready, invoke_ctx) in append_res {
             let region_id = invoke_ctx.region_id;
-            let res = self.region_peers
-                .get_mut(&region_id)
-                .unwrap()
-                .post_raft_ready_append(
+            let mut is_merging;
+            let res = {
+                let peer = self.region_peers.get_mut(&region_id).unwrap();
+                is_merging = peer.pending_merge.is_some();
+                peer.post_raft_ready_append(
                     &mut self.raft_metrics,
                     &self.trans,
                     &mut ready,
                     invoke_ctx,
-                );
+                )
+            };
+            if is_merging && res.is_some() {
+                // After applying a snapshot, PreMerge is rollback implicitly.
+                self.on_ready_rollback_pre_merge(region_id, 0, None);
```

Comment:
Why pre merge has to be rollbacked by applying snapshots?

Reply:
Because follower can't receive a snapshot until it's far behind leader, which indicates that `PreMerge` was rollback by leader already, otherwise the log gap can't increase.

Reply:
Got it. The follower can lost the `rollback` proposal.